File: schemas/paper.py
from __future__ import annotations

from datetime import date
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, HttpUrl


class AuthorDTO(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    orcid: str | None = None
    affiliation: str | None = None


class PaperDTO(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str
    authors: list[AuthorDTO] = Field(default_factory=list)
    abstract: str | None = None
    year: int | None = None
    publication_date: date | None = None
    doi: str | None = None
    arxiv_id: str | None = None
    pubmed_id: str | None = None
    source: str
    sources: list[str] = Field(default_factory=list)
    venue: str | None = None
    venue_issn: str | None = None
    publisher: str | None = None
    citation_count: int | None = None
    reference_count: int | None = None
    is_open_access: bool = False
    oa_url: HttpUrl | None = None
    landing_url: HttpUrl | None = None
    topics: list[str] = Field(default_factory=list)
    raw: dict[str, Any] = Field(default_factory=dict)

    relevance_score: float | None = None
    recency_score: float | None = None
    citation_score: float | None = None
    venue_score: float | None = None
    journal_quality_score: float | None = None
    open_access_score: float | None = None
    semantic_score: float | None = None
    is_predatory: bool = False
    quartile: str | None = None  # "Q1" | "Q2" | "Q3" | "Q4" — exposed for UI badges
    final_score: float | None = None

File: frontend/src/types/paper.ts
export interface AuthorDTO {
  name: string
  orcid?: string | null
  affiliation?: string | null
}

export interface PaperDTO {
  title: string
  authors: AuthorDTO[]
  abstract?: string | null
  year?: number | null
  publication_date?: string | null
  doi?: string | null
  arxiv_id?: string | null
  pubmed_id?: string | null
  source: string
  sources: string[]
  venue?: string | null
  venue_issn?: string | null
  publisher?: string | null
  citation_count?: number | null
  reference_count?: number | null
  is_open_access: boolean
  oa_url?: string | null
  landing_url?: string | null
  topics: string[]
  relevance_score?: number | null
  recency_score?: number | null
  citation_score?: number | null
  venue_score?: number | null
  journal_quality_score?: number | null
  open_access_score?: number | null
  semantic_score?: number | null
  is_predatory?: boolean
  quartile?: "Q1" | "Q2" | "Q3" | "Q4" | null
  final_score?: number | null
}

export interface SearchFilters {
  year_from?: number | null
  year_to?: number | null
  sources?: string[] | null
  open_access?: boolean | null
  min_citations?: number | null
  venue_contains?: string | null
  topic?: string | null
  institution?: string | null
  type?: string | null
}

export interface SearchRequest {
  query: string
  filters: SearchFilters
  limit: number
  offset?: number
  sort_by?: "relevance" | "year_desc" | "year_asc" | "citations"
}

export interface SearchResponse {
  total: number
  latency_ms: number
  results: PaperDTO[]
  offset: number
  has_more: boolean
  warnings?: string[]
}
